modules/tensor_mechanics/src/constraints/NodalStickConstraint.C

```diff
@@ -0,0 +1,297 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "NodalStickConstraint.h"
+#include "MooseMesh.h"
+#include "Assembly.h"
+#include "SystemBase.h"
+
+#include "libmesh/mesh_inserter_iterator.h"
+#include "libmesh/parallel.h"
+#include "libmesh/parallel_elem.h"
+#include "libmesh/parallel_node.h"
+
+// C++ includes
+#include <limits.h>
+
+registerMooseObject("MooseApp", NodalStickConstraint);
+
+template <>
+InputParameters
+validParams<NodalStickConstraint>()
+{
+  InputParameters params = validParams<NodalConstraint>();
+  params.addRequiredParam<BoundaryName>("boundary", "The master boundary");
+  params.addRequiredParam<BoundaryName>("slave", "The slave boundary");
+  params.addRequiredParam<Real>("penalty", "Stiffness of the spring.");
+  return params;
+}
+
+NodalStickConstraint::NodalStickConstraint(const InputParameters & parameters)
+  : NodalConstraint(parameters),
+    _master_node_set_id(getParam<BoundaryName>("boundary")),
+    _slave_node_set_id(getParam<BoundaryName>("slave")),
+    _penalty(getParam<Real>("penalty"))
+{
+  updateConstrainedNodes();
+}
+
+void
+NodalStickConstraint::meshChanged()
+{
+  updateConstrainedNodes();
+}
+
+void
+NodalStickConstraint::updateConstrainedNodes()
+{
+  _master_node_vector.clear();
+  _connected_nodes.clear();
+  _master_conn.clear();
+
+  std::vector<dof_id_type> slave_nodelist =
+      _mesh.getNodeList(_mesh.getBoundaryID(_slave_node_set_id));
+  std::vector<dof_id_type> master_nodelist =
+      _mesh.getNodeList(_mesh.getBoundaryID(_master_node_set_id));
+
+  std::vector<dof_id_type>::iterator in;
+
+  // Fill in _connected_nodes, which defines slave nodes in the base class
+  for (in = slave_nodelist.begin(); in != slave_nodelist.end(); ++in)
+  {
+    if (_mesh.nodeRef(*in).processor_id() == _subproblem.processor_id())
+      _connected_nodes.push_back(*in);
+  }
+
+  // Fill in _master_node_vector, which defines slave nodes in the base class
+  for (in = master_nodelist.begin(); in != master_nodelist.end(); ++in)
+    _master_node_vector.push_back(*in);
+
+  const auto & node_to_elem_map = _mesh.nodeToElemMap();
+  std::vector<std::vector<dof_id_type>> elems(_master_node_vector.size());
+
+  // Add elements connected to master node to Ghosted Elements.
+
+  // On a distributed mesh, these elements might have already been
+  // remoted, in which case we need to gather them back first.
+  if (!_mesh.getMesh().is_serial())
+  {
+    std::set<Elem *, CompareElemsByLevel> master_elems_to_ghost;
```

Comment:
The non-unity build error is because `CompareElemsByLevel` is defined in `NodalFrictionalConstraint.h`, which is not #included in this class. It works for the unity build because the headers all get combined together. You could fix it by adding `#include "NodalFricitonalConstraint.h" in this file, but that would be kind of ugly because it pulls in a lot of irrelevant stuff.  Probably best to move that into a separate file where utility stuff used by both of these types of constraints lives.

Reply:
Ok. I'll try putting the namespace stuff into a NodalCommonConstraint.h file and include that in both of the other classes. Thanks.

Reply:
Yeah, do something like that. I'd call it NodalConstraintUtils, though

Reply:
I'll update the name of the file and push it again. I think the fails in the testing are unrelated and should be fixed soon.